### PR TITLE
Plane:change Mode Takeoff to climb to TKOFF_ALT minimum if already flying on entry

### PR DESCRIPTION
**Pull request by Hwurzburg**
alternate to #23614

plus some minor comment corrections and code clean-ups

removed unnecessary code in TKOFF_LVL_ALT segment that distorts the target waypoint distance .......currently extends the target loiter waypoint away by the distance travelled during TKOFF_LVL period...

fixes another bug in that entering the mode on the ground before GPS lock(ie valid position) creates a bogus loiter position target...luckily, the TKOFF_LVL_ALT code resets the position (but incorrectly as discussed above).

This will prevent accidentally entering a loiter below TKOFF_ALT but also prevents using this mode in place of LOITER if you took off and then flew into a canyon and wanted to LOITER below your takeoff point altitude. I think think that the safety improvement is worth the corner case use restriction.

### CODE REVIEW

**Comment by timtuxworth** [comment]
I agree with @IamPete1 ` I think I prefer ... to always climb. Maybe just climb until the set height above ground rather than always climbing by the set distance.` Perhaps it should climb to _at least_ TKOFF_ALT rather than current alt + TKOFF_ALT, but should not descend if already at or above TKOFF_ALT.

But otherwise this one is definitely better IMO.

File: ArduPlane/mode_takeoff.cpp [hunk 1]
```diff
@@ -68,59 +68,78 @@ bool ModeTakeoff::_enter()
 
 void ModeTakeoff::update()
 {
+    // nothing works if we dont have a valid position!
+    if (!plane.current_loc.initialised()) {
+        return;
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:73]: Still need to do something. Either:
```
        plane.calc_nav_roll();
        plane.calc_nav_pitch();	
        plane.calc_throttle();
```
Or set targets to 0.
```
nav_roll_cd = 0;
nav_pitch_cd = 0;
plane.calc_throttle();
```

If this is basically a no-op with no postilion having position should be a requirement checked in the `_enter` function.

File: ArduPlane/mode_takeoff.cpp [hunk 2]
```diff
@@ -68,59 +68,78 @@ bool ModeTakeoff::_enter()
...
+    const float dist = target_dist;
+    const float alt = target_alt;  
     if (!takeoff_started) {
-        // see if we will skip takeoff as already flying
+        const float direction = degrees(ahrs.yaw);
+        // if you are flying but had not used Mode Takeoff before
+        // then start_loc not set and use home altitude as start altitude
+        int32_t alt_cm;
+        if (start_loc.alt == 0 && plane.current_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, alt_cm)) {
+            start_loc.alt = plane.current_loc.alt - alt_cm;
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:84]: I don't understand this. Your trying to get the absolute altitude of home? You could do:
`start_loc.set_alt_cm(0, Location::AltFrame::ABOVE_HOME);`

File: ArduPlane/mode_takeoff.cpp [hunk 3]
```diff
@@ -68,59 +68,78 @@ bool ModeTakeoff::_enter()
...
+        }
+        // if already flying skip takeoff climb if above TKOFF_ALT, otherwise climb then loiter
+        // if not flying setup target location TKOFF_DIST from
+        // takeoff point, at a height of TKOFF_ALT
         if (plane.is_flying() && (millis() - plane.started_flying_ms > 10000U) && ahrs.groundspeed() > 3) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Takeoff skipped - circling");
+            if (plane.current_loc.alt >= (start_loc.alt + alt*100.0)) {
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:90]: Should use the location methods to insure the frame is the same. EG:
```
float alt_diff;
if (plane.current_loc.get_alt_distance(start_loc, alt_diff) && (alt_diff > alt)) {
```

File: ArduPlane/mode_takeoff.cpp (hunk 1 above)
Comment (Hwurzburg) [reply ArduPlane/mode_takeoff.cpp]: I guess if you lose position while flying you still want to update roll and pitch...so I just will enclose the waypoint setup stuff with the qualifier...

File: ArduPlane/mode_takeoff.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/mode_takeoff.cpp]: yes trying to set an uninitialized start_loc to home alt so the following conditionals work....does the suggested line do the same? don't need the lat/long to be correct...

File: ArduPlane/mode_takeoff.cpp (hunk 3 above)
Comment (Hwurzburg) [reply ArduPlane/mode_takeoff.cpp]: why would the frame be different? these are locations and using their ASL alts

File: ArduPlane/mode_takeoff.cpp (hunk 2 above)
Comment (Hwurzburg) [reply ArduPlane/mode_takeoff.cpp]: got it set to home directly  and put it a check to assure there was a home

File: ArduPlane/mode_takeoff.cpp (hunk 3 above)
Comment (IamPete1) [reply ArduPlane/mode_takeoff.cpp]: They shouldn't be, but in general we do a terrible job of altitudes and keeping track of frames in plane. So if we can do it properly were less likely to mess it up in the future.

File: ArduPlane/mode_takeoff.cpp [hunk 4]
```diff
@@ -68,66 +68,80 @@ bool ModeTakeoff::_enter()
...
+        const float alt = target_alt;  
+        if (!takeoff_started) {
+            const float direction = degrees(ahrs.yaw);
+            // if you are flying but had not used Mode Takeoff before
+            // then start_loc not set and use home as start location
+            start_loc = AP::ahrs().get_home();
+            // if already flying skip takeoff climb if above TKOFF_ALT, otherwise climb then loiter
+            // if not flying setup target location TKOFF_DIST from
+            // takeoff point, at a height of TKOFF_ALT
+            if (plane.is_flying() && (millis() - plane.started_flying_ms > 10000U) && ahrs.groundspeed() > 3) {
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:84]: Should probably check throttle suppression too. `!plane.throttle_suppressed`

File: ArduPlane/mode_takeoff.cpp [hunk 5]
```diff
@@ -68,66 +68,80 @@ bool ModeTakeoff::_enter()
...
-        // or with poor initial compass
-        float direction = start_loc.get_bearing_to(plane.current_loc) * 0.01;
-        float dist_done = start_loc.get_distance(plane.current_loc);
         const float dist = target_dist;
+        const float alt = target_alt;  
+        if (!takeoff_started) {
+            const float direction = degrees(ahrs.yaw);
+            // if you are flying but had not used Mode Takeoff before
+            // then start_loc not set and use home as start location
+            start_loc = AP::ahrs().get_home();
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:80]: This bit should be in the "is flying" if statement.

Comment (IamPete1) [reply ArduPlane/mode_takeoff.cpp:80]: using home is a bit odd. I think you really want current location with home altitude? You we do a similar thing in QRTL to climb first, we use terrain alt there:

https://github.com/ArduPilot/ardupilot/blob/4a8eda058997548fd8979848f450b7e831274fa9/ArduPlane/mode_qrtl.cpp#L38-L48

**Comment by IamPete1** [comment]
I think this is a good idea, maybe the diff is making it look more complicated than it is but I would have expected a smaller code change, only in the is flying check.

**Review by tridge — CHANGES_REQUESTED** [review]
should use relative_ground_altitude() function for height above ground
should not descend

**Comment by Hwurzburg** [comment]
change requested is done

**Comment by Hwurzburg** [comment]
closing in favor of #23728 which has an unsquashed commit tree for easy reviewing of each change made to the file

**Pull request closed without merging**